File: Hangman Equivalency/Hangman.java
package hangman;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

/**
 * @author Samuel Gardner
 */
public class Hangman
{
	public static void clearConsole()
	{
		//Copied from: https://stackoverflow.com/questions/2979383/java-clear-the-console
		if (System.getProperty("os.name").contains("Windows"))
		{
			try
			{
				new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
			}
			catch (InterruptedException | IOException e)
			{
				e.printStackTrace();
			}
		}
		else
		{
			System.out.println("\n\n"); //for mac/linux
		}
	}

	public static void main(String[] args)
	{
		final String alphaAndSpaceReg = "^[a-zA-Z ]+$";
		final String alphaCharReg = "[a-zA-Z]";
		final String affirmativeReg = "^[Y|yE|eS|s]|[Y|y]";
		final String negativeReg = "^[N|nO|o]|[N|n]|[Q|qU|uI|iT|t]";
		try (Scanner sc = new Scanner(System.in))
		{
			boolean running = true;
			while (running)
			{
				clearConsole();
				boolean gaming = true;
				boolean validInput = false;
				String fullWord = "";
				StringBuilder guessWord = new StringBuilder();
				ArrayList<String> guessedLetters = new ArrayList<String>();
				int livesLeft = 6;
				System.out.println("Welcome to hangman! Enter a word to start: (Can contain only letters and spaces.) ");
				while (!validInput)
				{
					// Cleans the input.
					List<String> l = Arrays.asList(sc.nextLine().strip().split(" "));
					l.removeIf(x -> x.isBlank());
					fullWord = String.join(" ", l);
					if (fullWord.matches(alphaAndSpaceReg))
					{
						validInput = true;
					}
					else
					{
						System.out.println("Sorry, that input isn't valid. You can only choose a word or phrase that contains letters and spaces.");
						System.out.println("Enter a word to start: (Can contain only letters and spaces.) ");
					}
				}
				// We now have a valid input.
				guessWord.append(fullWord.replaceAll(alphaCharReg, "_"));
				clearConsole();
				// print status bar:
				System.out.println("Lives Left: " + livesLeft + ".   Letters guessed: " + String.join(", ", guessedLetters) + "\n");
				// print word:
				System.out.println(guessWord.toString());
				while (gaming)
				{
					// print prompt:
					boolean validGuess = false;
					char guess = '\0';
					while (!validGuess)
					{
						System.out.println("\nGuess a letter: ");
						String s = sc.nextLine().toLowerCase().strip();
						guess = s.length() > 0 ? s.charAt(0) : '\0';
						if (guess == '\0')
						{
							System.out.println("Make sure to type a character. The first character entered will be your guess.");
						}
						else if (guessedLetters.contains(Character.toString(guess)))
						{
							System.out.println("You already guessed that letter!");
						}
						else if (!Character.toString(guess).matches(alphaCharReg))
						{
							System.out.println("You can only guess English letters!");
						}
						else
						{
							validGuess = true;
						}
					}
					guessedLetters.add(Character.toString(guess));
					// evaluate:
					boolean didFindLetter = false;
					for (int i = 0; i < fullWord.length(); i++)
					{
						if (fullWord.toLowerCase().charAt(i) == guess)
						{
							didFindLetter = true;
							guessWord.replace(i, i + 1, "");
							if (i == 0 || fullWord.charAt(i - 1) == ' ')
							{
								guessWord.insert(i, Character.toString(guess).toUpperCase());
							}
							else
							{
								guessWord.insert(i, Character.toString(guess).toLowerCase());
							}
						}
					}
					if (!didFindLetter)
					{
						livesLeft--;
					}
					// Clear for fresh start.
					clearConsole();
					// print status bar:
					System.out.println("Lives Left: " + livesLeft + ".   Letters guessed: " + String.join(", ", guessedLetters) + "\n");
					// print word:
					System.out.println(guessWord.toString());
					// check for win/lose.
					if (livesLeft == 0 || !guessWord.toString().contains("_"))
					{
						if (livesLeft == 0)
						{
							//You lose.
							System.out.println("Uh oh! Out of lives, you lose. The word/phrase was: \"" + fullWord + "\" Play Again? (Y/N)");
						}
						else
						{
							//You win!
							System.out.println("Congratulations! You win! Play Again? (Y/N)");
						}
						String input = "";
						validInput = false;
						while (!validInput)
						{
							input = sc.nextLine();
							if (input.matches(affirmativeReg) || input.matches(negativeReg))
							{
								validInput = true;
							}
							else
							{
								System.out.println("Hey! That isn't a valid input, say \"Y\" or \"N\".");
							}
						}
						if (input.matches(affirmativeReg))
						{
							livesLeft = 6;
							guessedLetters.clear();
							guessWord.setLength(0);
							fullWord = "";
							gaming = false;
							running = true;
						}
						else if (input.matches(negativeReg))
						{
							System.out.println("Thanks for playing!");
							try
							{
								Thread.sleep(2000);
							}
							catch (InterruptedException e)
							{
								e.printStackTrace();
							}
							gaming = false;
							running = false;
						}
					}
				}
			}
		}
	}
}

File: Hangman Equivalency/Hangman.py
import re
import os
import time

def clearConsole(): 
   os.system('cls' if os.name == 'nt' else 'clear')
   #print("\n\n\n") #it's a royal pain to clear the console in python apparently so just add three newlines instead.

alphaAndSpaceReg = re.compile("^[a-zA-Z ]+$")
alphaCharReg = re.compile("[a-zA-Z]")
affirmativeReg = re.compile("^[Y|yE|eS|s]|[Y|y]")
negativeReg = re.compile("^[N|nO|o]|[N|n]|[Q|qU|uI|iT|t]")

running: bool = True
while running:
    clearConsole()
    gaming: bool = True
    validInput: bool = False
    fullWord: str = ""
    guessWord: str = ""
    guessedLetters = []
    livesLeft: int = 6
    print("Welcome to hangman! Enter a word to start: (Can contain only letters and spaces.) ")
    while not validInput:
        #cleans the input
        fullWord = " ".join(input().strip().lower().split())
        if alphaAndSpaceReg.fullmatch(fullWord):
            validInput = True
        else:
            print("Sorry, that input isn't valid. You can only choose a word or phrase that contains letters and spaces.")
            print("Enter a word to start: (Can contain only letters and spaces.) ")
    #we now have valid input
    for x in fullWord:
        if alphaCharReg.match(x):
            guessWord += "_"
        else:
            guessWord += " "
    clearConsole()
    #print status bar
    print("Lives Left: " + str(livesLeft) + ".   Letters guessed: " + ", ".join(guessedLetters) + "\n")
    #print word:
    print(guessWord)
    while gaming:
        #print prompt
        validGuess: bool = False
        guess = " "
        while not validGuess:
            print("\nGuess a letter: ")
            guess = input().strip().lower()
            guess = guess[0] if len(guess) > 0 else " "
            if guess == " ":
                print("Make sure to type a character. The first character entered will be your guess.")
            elif guess in guessedLetters:
                print("You already guessed that letter!")
            elif not alphaCharReg.fullmatch(guess):
                print("You can only guess English letters!")
            else:
                validGuess = True
        guessedLetters.append(guess)
        #evaluate
        didFindLetter: bool = False
        for i, val in enumerate(fullWord):
            if val.lower() == guess:
                didFindLetter = True
                if i == 0 or fullWord[i - 1] == " ":
                    guessWord = guessWord[:i] + guess.upper() + guessWord[i + 1:]
                else:
                    guessWord = guessWord[:i] + guess.lower() + guessWord[i + 1:]
        if not didFindLetter:
            livesLeft -= 1
        #Clear for fresh start
        clearConsole()
        #print status bar:
        print("Lives Left: " + str(livesLeft) + ".   Letters guessed: " + ", ".join(guessedLetters) + "\n")
        #print word
        print(guessWord)
        #check for win/lose
        if livesLeft == 0 or guessWord.find("_") == -1:
            if livesLeft == 0:
                print("Uh oh! Out of lives, you lose. The word/phrase was: \"" + fullWord + "\" Play Again? (Y/N)")
            else:
                print("Congratulations! You win! Play Again? (Y/N)")
            inputVal: str = ""
            validInput = False
            while not validInput:
                inputVal = input()
                if affirmativeReg.fullmatch(inputVal) or negativeReg.fullmatch(inputVal):
                    validInput = True
                else:
                    print("Hey! That isn't a valid input, say \"Y\" or \"N\".")
            if affirmativeReg.fullmatch(inputVal):
                livesLeft = 6
                guessedLetters.clear()
                guessWord = ""
                fullWord = ""
                gaming = False
                running = True
            elif negativeReg.fullmatch(inputVal):
                print("Thanks for playing!")
                time.sleep(2)
                gaming = False
                running = False